cmd/bee-split/main.go

```diff
@@ -0,0 +1,241 @@
+}
+
+// fsStore provides a storage.Putter that writes chunks directly to the filesystem.
+// Each chunk is a separate file, where the hex address of the chunk is the file name.
+type fsStore struct {
+	path string
+}
+
+// newFsStore creates a new fsStore
+func newFsStore(path string) storage.Putter {
+	return &fsStore{
+		path: path,
+	}
+}
+
+// Put implements storage.Putter
+func (f *fsStore) Put(ctx context.Context, mode storage.ModePut, chs ...swarm.Chunk) (exist []bool, err error) {
+	for _, ch := range chs {
+		chunkPath := filepath.Join(f.path, ch.Address().String())
+		err := ioutil.WriteFile(chunkPath, ch.Data(), 0o666)
+		if err != nil {
+			return []bool{}, err
+		}
+	}
+	return []bool{}, nil
+}
+
+// apiStore provies a storage.Putter that adds chunks to swarm through the HTTP chunk API.
+type apiStore struct {
+	baseUrl string
+}
+
+// newApiStore creates a new apiStor
+func newApiStore(host string, port int, ssl bool) (storage.Putter, error) {
+	scheme := "http"
+	if ssl {
+		scheme += "s"
+	}
+	u := &url.URL{
+		Host:   fmt.Sprintf("%s:%d", host, port),
+		Scheme: scheme,
+		Path:   "bzz-chunk",
+	}
+	return &apiStore{
+		baseUrl: u.String(),
+	}, nil
+}
+
+// Put implements storage.Putter
+func (a *apiStore) Put(ctx context.Context, mode storage.ModePut, chs ...swarm.Chunk) (exist []bool, err error) {
+	for _, ch := range chs {
+		addr := ch.Address().String()
+		buf := bytes.NewReader(ch.Data())
+		url := strings.Join([]string{a.baseUrl, addr}, "/")
+		c := &http.Client{}
+		res, err := c.Post(url, "application/octet-stream", buf)
+		if err != nil {
+			return []bool{}, err
+		}
+		_ = res
+	}
+	return []bool{}, nil
+}
+
+// limitReadCloser wraps the input to the application to limit the input to the given count flag.
+type limitReadCloser struct {
+	io.Reader
+	closeFunc func() error
+}
+
+// newLimitReadCloser creates a new limitReadCloser.
+func newLimitReadCloser(r io.Reader, closeFunc func() error, c int64) io.ReadCloser {
+	return &limitReadCloser{
+		Reader:    io.LimitReader(r, c),
+		closeFunc: closeFunc,
+	}
+}
+
+// Close implements io.Closer
+func (l *limitReadCloser) Close() error {
+	return l.closeFunc()
+}
+
+// Split is the underlying procedure for the CLI command
+func Split(cmd *cobra.Command, args []string) (err error) {
+	var infile io.ReadCloser
+
+	// if one arg is set, this is the input file
+	// if not, we are reading from standard input
+	if len(args) > 0 {
+
+		// get the file length
+		info, err := os.Stat(args[0])
+		if err != nil {
+			return err
+		}
+		fileLength := info.Size()
+
+		// check if we are limiting the input, and if the limit is valid
+		if inputLength > 0 {
+			if inputLength > fileLength {
+				return fmt.Errorf("input data length set to %d on file with length %d", inputLength, fileLength)
+			}
+		} else {
+			inputLength = fileLength
+		}
+
+		// open file and wrap in limiter
+		f, err := os.Open(args[0])
+		if err != nil {
+			return err
+		}
+		defer f.Close()
```

Comment:
this is strange to me and i think this is a potential panic. you're calling defer (this will be executed only when the entire `Split` functions runs to end, but you're providing `f.Close` to the `ReadCloser`, which means some other calling context might try to close the file twice. can you elaborate on why is this needed in this manner?

Reply:
> which means some other calling context might try to close the file twice

what do you mean? `os.Open` opens a unique file descriptor.

Reply:
since you're passing `f.Close` to the `ReadCloser` it means some other context can call the close, and it will cause the `defer` statement to panic since the file handle is already closed.